pkg/kubelet/images/pullmanager/image_pull_policies.go

```diff
@@ -0,0 +1,151 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pullmanager
+
+import (
+	"fmt"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+	kubeletconfiginternal "k8s.io/kubernetes/pkg/kubelet/apis/config"
+)
+
+// ImagePullPolicyEnforcer defines a class of functions implementing a credential
+// verification policies for image pulls. These function determines whether the
+// implemented policy requires credential verification based on image name, local
+// image presence and existence of records about previous image pulls.
+//
+// `image` is an image name from a Pod's container "image" field.
+// `imagePresent` informs whether the `image` is present on the node.
+// `imagePulledByKubelet` marks that ImagePulledRecord or ImagePullingIntent records
+// for the `image` exist on the node, meaning it was pulled by the kubelet somewhere
+// in the past.
+type ImagePullPolicyEnforcer func(image string, imagePresent, imagePulledByKubelet bool) bool
+
+func NewImagePullCredentialVerificationPolicy(policy kubeletconfiginternal.ImagePullCredentialsVerificationPolicy, imageAllowList []string) (ImagePullPolicyEnforcer, error) {
+	switch policy {
+	case kubeletconfiginternal.NeverVerify:
+		return ImagePullPolicyEnforcer(NeverVerifyImagePullPolicy), nil
+	case "", kubeletconfiginternal.NeverVerifyPreloadedImages:
+		return ImagePullPolicyEnforcer(NeverVerifyPreloadedPullPolicy), nil
+	case kubeletconfiginternal.NeverVerifyAllowlistedImages:
+		return NewNeverVerifyAllowListedPullPolicy(imageAllowList)
+	case kubeletconfiginternal.AlwaysVerify:
+		return ImagePullPolicyEnforcer(AlwaysVerifyImagePullPolicy), nil
+	default:
+		return nil, fmt.Errorf("unknown image pull credential verification policy: %v", policy)
+	}
+}
+
+func NeverVerifyImagePullPolicy(image string, imagePresent, imagePulledByKubelet bool) bool {
+	return false
+}
+
+func NeverVerifyPreloadedPullPolicy(image string, imagePresent, imagePulledByKubelet bool) bool {
+	if imagePresent && !imagePulledByKubelet {
+		return false
+	}
+	return true
+}
+
+func AlwaysVerifyImagePullPolicy(image string, imagePresent, imagePulledByKubelet bool) bool {
+	return true
+}
+
+type NeverVerifyAllowlistedImages struct {
+	absoluteURLs sets.Set[string]
+	prefixes     []string
+}
+
+func (p *NeverVerifyAllowlistedImages) RequiresVerification(image string, imagePresent, imagePulledByKubelet bool) bool {
+	if imagePresent && p.imageMatches(image) {
+		return false
+	}
+	return true
+}
+
+func (p *NeverVerifyAllowlistedImages) imageMatches(image string) bool {
+	if p.absoluteURLs.Has(image) {
+		return true
+	}
+	for _, prefix := range p.prefixes {
+		if strings.HasPrefix(image, prefix) {
+			return true
+		}
+	}
+	return false
+}
+
+func (p *NeverVerifyAllowlistedImages) AllowPattern(pattern string) error {
```

Comment:
I would probably inline this into the constructor.. AllowPattern is confusing to export and sounds like RequiresVerification